Source/Scene/Cesium3DTileBatchTable.js

```diff
@@ -0,0 +1,1519 @@
+            var binaryProperties = Cesium3DTileBatchTable.getBinaryProperties(classInstancesLength, properties, binary);
+            classes[i].instances = combine(binaryProperties, properties);
+        }
+
+        var classCounts = arrayFill(new Array(classesLength), 0);
+        var classIndexes = new Uint16Array(instancesLength);
+        for (i = 0; i < instancesLength; ++i) {
+            classId = classIds[i];
+            classIndexes[i] = classCounts[classId];
+            ++classCounts[classId];
+        }
+
+        var hierarchy = {
+            classes : classes,
+            classIds : classIds,
+            classIndexes : classIndexes,
+            parentCounts : parentCounts,
+            parentIndexes : parentIndexes,
+            parentIds : parentIds
+        };
+
+        //>>includeStart('debug', pragmas.debug);
+        validateHierarchy(hierarchy);
+        //>>includeEnd('debug');
+
+        return hierarchy;
+    }
+
+    //>>includeStart('debug', pragmas.debug);
+    var scratchValidateStack = [];
+    function validateHierarchy(hierarchy) {
+        var stack = scratchValidateStack;
+        stack.length = 0;
+
+        var classIds = hierarchy.classIds;
+        var instancesLength = classIds.length;
+
+        for (var i = 0; i < instancesLength; ++i) {
+            validateInstance(hierarchy, i, stack);
+        }
+    }
+
+    function validateInstance(hierarchy, instanceIndex, stack) {
+        var parentCounts = hierarchy.parentCounts;
+        var parentIds = hierarchy.parentIds;
+        var parentIndexes = hierarchy.parentIndexes;
+        var classIds = hierarchy.classIds;
+        var instancesLength = classIds.length;
+
+        if (!defined(parentIds)) {
+            // No need to validate if there are no parents
+            return;
+        }
+
+        if (instanceIndex >= instancesLength) {
+            throw new DeveloperError('Parent index ' + instanceIndex + ' exceeds the total number of instances: ' + instancesLength);
+        }
+        if (stack.indexOf(instanceIndex) > -1) {
+            throw new DeveloperError('Circular dependency detected in the batch table hierarchy.');
+        }
+
+        stack.push(instanceIndex);
+        var parentCount = defined(parentCounts) ? parentCounts[instanceIndex] : 1;
+        var parentIndex = defined(parentCounts) ? parentIndexes[instanceIndex] : instanceIndex;
+        for (var i = 0; i < parentCount; ++i) {
+            var parentId = parentIds[parentIndex + i];
+            // Stop the traversal when the instance has no parent (its parentId equals itself), else continue the traversal.
+            if (parentId !== instanceIndex) {
+                validateInstance(hierarchy, parentId, stack);
+            }
+        }
+        stack.pop(instanceIndex);
+    }
+    //>>includeEnd('debug');
+
+    Cesium3DTileBatchTable.getBinaryProperties = function(featuresLength, json, binary) {
+        var binaryProperties;
+        if (defined(json)) {
```

Comment:
I believe this check is superflous, `for(var name in undefined)` and `for(var name in null)` will both result in zero iteration.